### PR TITLE
Pass missing infer parameters during conversion

### LINKED ISSUE #3365: Python SDK: Optional `InferInput.parameters` missing from `InferRequest.to_rest()/.to_grpc()`
/kind bug

**What steps did you take and what happened:**
Hi! In attempting to pass `InferInput.parameters` from a transformer to a predictor I realised the request reaching the predictor had completely removed the parameters.

This can be replicated with:

```python
import numpy as np
from kserve import InferRequest, InferInput

test_array = np.zeros((2, 2, 1), dtype="uint8")

test_input = InferInput(name="test", shape=test_array.shape, datatype="UINT8", parameters={"hello": "world"})
test_input.set_data_from_numpy(test_array)

test_infer = InferRequest(model_name="test", infer_inputs=[test_input])

print(test_infer.inputs[0].parameters)

test_infer.to_rest()
```

Output:
```terminal
{'hello': 'world', 'binary_data_size': 4}

{'id': None,
 'inputs': [{'name': 'test',
   'shape': (2, 2, 1),
   'datatype': 'UINT8',
   'data': None}]}
```

And is expected from looking at the code as the parameters are not handled:

https://github.com/kserve/kserve/blob/c52753519e1802e73581856b1ae3a0f2f178d457/python/kserve/kserve/protocol/infer_type.py#L271-L329

Which is called by the `_X_predict` functions in the `kserve.Model` class.

**What did you expect to happen:**
I expect the parameters to be preserved when using the `to_rest/grpc` methods.

Am I mis-understanding something about parameters in that this is expected behaviour?

**What's the InferenceService yaml:**

N/A

**Anything else you would like to add:**

Would appear to be the same for `InferOutput`.

**Environment:**

- KServe Version: 0.11.2

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3365

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Pass missing infer parameters during conversion
```

### CODE REVIEW

**Comment by ajstewart** [comment]
I saw this PR linked to my issue and had a look out of interest as I wondered about the `to_grpc` method. In my own work around I had to make sure that the `to_grpc` was setting the parameters as `kserve.protocol.grpc.grpc_predict_v2_pb2.InferParameter` objects. This is in the transformer. Then in the predictor I had to do some conversion back to either string, int, float or bool with prior knowledge as to what they should be (I couldn't see a helper on that object to transform back to the intended format).

This is all new to me and I'm not very familiar with kserve so I might be talking nonsense but I couldn't see an elegant way to deal with that.

I must admit I'm using them to place extra information into the request to change some predictor behaviour and some custom component settings, so I could be abusing the functionality.

**Comment by sivanantha321** [comment]
> I saw this PR linked to my issue and had a look out of interest as I wondered about the `to_grpc` method. In my own work around I had to make sure that the `to_grpc` was setting the parameters as `kserve.protocol.grpc.grpc_predict_v2_pb2.InferParameter` objects. This is in the transformer. Then in the predictor I had to do some conversion back to either string, int, float or bool with prior knowledge as to what they should be (I couldn't see a helper on that object to transform back to the intended format).
> 
> This is all new to me and I'm not very familiar with kserve so I might be talking nonsense but I couldn't see an elegant way to deal with that.
> 
> I must admit I'm using them to place extra information into the request to change some predictor behaviour and some custom component settings, so I could be abusing the functionality.

Can you try with these changes and give some feedback ?

File: python/kserve/kserve/protocol/infer_type.py [hunk 1]
```diff
@@ -630,4 +646,44 @@ def to_grpc(self) -> ModelInferResponse:
...
+    :return: converted parameters as Dict[str, InferParameter]
+    :raises InvalidInput: if the parameter type is not supported.
+    """
+    if isinstance(parameters, dict):
+        for key, val in parameters.items():
+            if isinstance(val, str):
+                parameters[key] = InferParameter(string_param=val)
+            elif isinstance(val, bool):
+                parameters[key] = InferParameter(bool_param=val)
+            elif isinstance(val, int):
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/infer_type.py]: We do not support float ?

**Comment by yuzisun** [comment]
The change looks good, discussed with @siva to add a few tests

File: python/kserve/kserve/protocol/infer_type.py (hunk 1 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/infer_type.py]: Yes, we don't support float. https://github.com/kserve/kserve/blob/c254e704c3719d3310391ee994123d5b49588e5d/python/kserve/kserve/protocol/grpc/grpc_predict_v2.proto#L269C1-L284C2

**Comment by sivanantha321** [comment]
> The change looks good, discussed with @Siva to add a few tests

Tests added.

File: python/kserve/test/test_server.py [hunk 2]
```diff
@@ -240,14 +243,14 @@ def test_list_models_v1(self, http_server_client):
 
     def test_predict_v1(self, http_server_client):
         resp = http_server_client.post('/v1/models/TestModel:predict',
-                                       data=b'{"instances":[[1,2]]}')
+                                       content=b'{"instances":[[1,2]]}')
```
Comment (sivanantha321) [inline python/kserve/test/test_server.py:245]: To fix the warning
DeprecationWarning: Use 'content=<...>' to upload raw bytes/text content.

**Comment by ajstewart** [comment]
> Can you try with these changes and give some feedback ?

A gave your branch a go (https://github.com/kserve/kserve/pull/3368/commits/3a773e73a10d0257a786a6314a63e102a4a62891) but I'm hitting an assignment error pasted below. I think I remember hitting the same problem in my workaround and I ended not trying to reassign on the current parameters but instead recreating them.

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 276, in __call__
    await super().__call__(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/usr/local/lib/python3.10/site-packages/timing_asgi/middleware.py", line 68, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 237, in app
    raw_response = await run_endpoint_function(
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 163, in run_endpoint_function
    return await dependant.call(**values)
  File "/usr/local/lib/python3.10/site-packages/kserve/protocol/rest/v2_endpoints.py", line 151, in infer
    response, response_headers = self.dataplane.encode(model_name=model_name,
  File "/usr/local/lib/python3.10/site-packages/kserve/protocol/dataplane.py", line 273, in encode
    response = response.to_rest()
  File "/usr/local/lib/python3.10/site-packages/kserve/protocol/infer_type.py", line 611, in to_rest
    res['parameters'] = to_http_parameters(self.parameters)
  File "/usr/local/lib/python3.10/site-packages/kserve/protocol/infer_type.py", line 688, in to_http_parameters
    parameters[key] = val.string_param
ValueError: Direct assignment of submessage not allowed
```

File: python/kserve/test/test_infer_type.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                                                         "test-bool": InferParameter(bool_param=True),
+                                                         "test-int": InferParameter(int64_param=100)
+                                                     })],
+                                     from_grpc=True
+                                     )
+            res = InferResponse.from_grpc(infer_res)
+            assert InferResponseMatcher(expected).__eq__(res) is True
+
+
+class InferInputMatcher:
```
Comment (yuzisun) [inline python/kserve/test/test_infer_type.py]: I think these can implemented in InferInput type itself

File: python/kserve/test/test_infer_type.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+            return False
+        if len(self.req.inputs) != len(other.inputs):
+            return False
+        for i in range(len(self.req.inputs)):
+            if not InferInputMatcher(self.req.inputs[i]).__eq__(other.inputs[i]):
+                return False
+        return True
+
+
+class InferOutputMatcher:
```
Comment (yuzisun) [inline python/kserve/test/test_infer_type.py]: Same here should be defined in InferOutput

File: python/kserve/test/test_infer_type.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                                         InferOutput(name="output-0", datatype="INT32", shape=[1, 2], data=[1, 2],
+                                                     parameters={
+                                                         "test-str": InferParameter(string_param="dummy"),
+                                                         "test-bool": InferParameter(bool_param=True),
+                                                         "test-int": InferParameter(int64_param=100)
+                                                     })],
+                                     from_grpc=True
+                                     )
+            res = InferResponse.from_grpc(infer_res)
+            assert InferResponseMatcher(expected).__eq__(res) is True
```
Comment (yuzisun) [inline python/kserve/test/test_infer_type.py]: You should be able to use operator == after __eq__ is defined

Comment (yuzisun) [reply python/kserve/test/test_infer_type.py]: Expected == res

File: python/kserve/test/test_infer_type.py (hunk 3 above)
Comment (sivanantha321) [reply python/kserve/test/test_infer_type.py]: done

File: python/kserve/test/test_infer_type.py (hunk 4 above)
Comment (sivanantha321) [reply python/kserve/test/test_infer_type.py]: done

File: python/kserve/test/test_infer_type.py (hunk 5 above)
Comment (sivanantha321) [reply python/kserve/test/test_infer_type.py]: done

**Comment by sivanantha321** [comment]
> > Can you try with these changes and give some feedback ?
> 
> A gave your branch a go ([3a773e7](https://github.com/kserve/kserve/commit/3a773e73a10d0257a786a6314a63e102a4a62891)) but I'm hitting an assignment error pasted below. I think I remember hitting the same problem in my workaround and I ended not trying to reassign on the current parameters but instead recreating them.
> 
> ```python
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
>     result = await app(  # type: ignore[func-returns-value]
>   File "/usr/local/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
>     return await self.app(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 276, in __call__
>     await super().__call__(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/starlette/applications.py", line 122, in __call__
>     await self.middleware_stack(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
>     raise exc
>   File "/usr/local/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
>     await self.app(scope, receive, _send)
>   File "/usr/local/lib/python3.10/site-packages/timing_asgi/middleware.py", line 68, in __call__
>     await self.app(scope, receive, send_wrapper)
>   File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
>     raise exc
>   File "/usr/local/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
>     await self.app(scope, receive, sender)
>   File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
>     raise e
>   File "/usr/local/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
>     await self.app(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
>     await route.handle(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
>     await self.app(scope, receive, send)
>   File "/usr/local/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
>     response = await func(request)
>   File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 237, in app
>     raw_response = await run_endpoint_function(
>   File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 163, in run_endpoint_function
>     return await dependant.call(**values)
>   File "/usr/local/lib/python3.10/site-packages/kserve/protocol/rest/v2_endpoints.py", line 151, in infer
>     response, response_headers = self.dataplane.encode(model_name=model_name,
>   File "/usr/local/lib/python3.10/site-packages/kserve/protocol/dataplane.py", line 273, in encode
>     response = response.to_rest()
>   File "/usr/local/lib/python3.10/site-packages/kserve/protocol/infer_type.py", line 611, in to_rest
>     res['parameters'] = to_http_parameters(self.parameters)
>   File "/usr/local/lib/python3.10/site-packages/kserve/protocol/infer_type.py", line 688, in to_http_parameters
>     parameters[key] = val.string_param
> ValueError: Direct assignment of submessage not allowed
> ```

I am not reassigning the parameters dict anymore. Can you retry with these changes ?

**Comment by ajstewart** [comment]
> I am not reassigning the parameters dict anymore. Can you retry with these changes ?

Yep these are being passed through now! The model I am testing is using grpc so once they get to the predictor I have to run the `to_http_parameters` so I can easily make use of the values in the dict of `InferParameter` types. But all the functionality is there now to do what I wanted to do, thanks!

**Comment by yuzisun** [comment]
Thanks @sivanantha321 fixing this!

/lgtm
/approve

**Pull request merged**